src/rmkit/fb/fb.cpy

```diff
@@ -211,6 +211,15 @@ namespace framebuffer:
             if (i + j) % 2 == 0 || (i + j) % 3 == 0:
               ptr[i+j*self.width] = WHITE
           break
+        case BLACK:
+           if dither != 1.0:
+             if dither > dither_noise(i, j):
+                ptr[i + j*self.width] = BLACK
+             else
+                ptr[i + j*self.width] = WHITE
+           else:
+             ptr[i + j*self.width] = color
+           break
```

Comment:
shouldn't this apply to white too? (or is that the case below?)

Reply:
Good question -- I started heading in that direction, but realized that color and dithering are basically the same thing.

I think the implementation for WHITE is the same as black, but 1-dither, i.e.
```cpp
if color == WHITE && dither != 1.0:
  return do_dithering(ptr, i, j, BLACK, 1.0 - dither)
```

I drew the samples by changing all my calls for filled shapes to use color=BLACK, dither=gray_level. I wonder if the thing to do is make the software dithering mode a property of FB, and then just use remarkable_color everywhere? So, maybe there's random dithering, bayer dithering, blue noise dithering, all of which draw only black and white pixels, and remarkable_color is used to set the threshold. And if you specifically set FB to _not_ use dithering, you get the "real" grayscale color rendered to the fb (and you can't use direct update).

Reply:
maybe can have DITHER_BAYER, DITHER_BLUE_NOISE, DITHER_RANDOM as color defines, similar to ERASER_STYLUS and ERASER_RUBBER? currently, it is remarkable_color = holding the type of fill pattern to use (have you seen what happens with ERASER_STYLUS/RUBBER?) and dither = amount to apply

as a consumer, if dither is a property on FB, i might expect draw_bitmap to also support dithering for fast update purposes.

Reply:
> So, maybe there's random dithering, bayer dithering, blue noise dithering, all of which draw only black and white pixels, and remarkable_color is used to set the threshold.

i am trying to understand how this works, because i thought dithering will have to do a pass over all the framebuffer (can you dither as you apply new content? i thought dithering will require neighboring pixel values when not painting a solid color)

Reply:
> as a consumer, if dither is a property on FB, i might expect draw_bitmap to also support dithering for fast update purposes.

We could definitely dither bitmaps. I took a shot at the demo image from libremarkable (I'm a little confused at how image loading works: it looks like image_data.buffer expects to be 4-channel rgba?)

<details>
<summary>This looks a little better on the remarkable than on a computer IMO</summary>

![blue_noise_parrot](https://user-images.githubusercontent.com/1356700/105509877-fbda3500-5c9b-11eb-9c01-7f9cb532c0fd.png)


</details>

Reply:
> i am trying to understand how this works, because i thought dithering will have to do a pass over all the framebuffer (can you dither as you apply new content? i thought dithering will require neighboring pixel values when not painting a solid color)

I'm pretty new to dithering, but as I understand it there are two main approaches used today. This is [ordered dithering](https://en.wikipedia.org/wiki/Ordered_dithering), which uses a matrix of different threshold values, and is completely deterministic: given x, y, and a color, you should get the exact same black-or-white return every time. There's also [error-diffusion](https://en.wikipedia.org/wiki/Error_diffusion), which I don't know as much about, but requires some state: for each pixel you pick black or white, but then use the error as input for the next pixel.

Reply:
> given x, y, and a color, you should get the exact same black-or-white return every time

I'm not 100% sure how this works if you were to say, dither but allow the full 16 shades of gray. I'm pretty sure the idea is pick the closest of the 16 shades, and then use the threshold matrix to adjust either up or down one shade.

Reply:
ok, putting dither type on the framebuffer sounds good to me, thanks for explaining. 

the original reason for the "dithering" was for the paintbrushes in harmony, it was really acting as a fill type, not as true dithering. if you open up harmony and draw a little bit with the pencil (will show random dithering), then try out the two different erasers, you will see that one eraser gives a partial erase effect (effectively lightening the area) until the stylus is let up, while the other draws a square fill pattern. is this orthogonal to dithering? or is dithering and fill type both under the same umbrella?

re: image.buffer: the buffer will contain the number of channels, so a grayscale image = 1 byte per pixel, rgb(a) is usually 4 byte. draw_bitmap will have logic for switching on number of channels. 
 
```cpp

// somewhere in fb.cpy
      int neww, newh, channels // an output parameter
      decoded := stbi_load(full_path, &neww, &newh, &channels, 4);
      image := image_data{(uint32_t*) decoded, (int) neww, (int) newh, channels}
      self->draw_bitmap(image,0,0,0)
      free(image.buffer)
```

Reply:
> is this orthogonal to dithering? or is dithering and fill type both under the same umbrella?

I think these are orthogonal -- maybe the thing to do is have do_dithering (or something like it) be responsible for coloring every pixel. Currently I think it only gets called via draw_rect and draw_line, but circles and bitmaps don't hit that path.

I like the patterns idea -- some of the puzzles are based on color and don't make a lot of sense without it (e.g. [flood](https://www.chiark.greenend.org.uk/~sgtatham/puzzles/js/flood.html), [map](https://www.chiark.greenend.org.uk/~sgtatham/puzzles/js/map.html), [same game](https://www.chiark.greenend.org.uk/~sgtatham/puzzles/js/samegame.html)). I've actually been thinking about reserving certain colors to mean "pattern". The backend uses floats where 0.0 is black and 1.0 is white, so I was considering 2.0 to be "hatched", and 3.0 to be "dotted", or whatever. That's probably getting ahead of myself a good bit since I haven't tried implementing those games yet.